ICE: `Foo was a subtype of &Foo but now is not?`

<!--
[31mICE[0m: Rustc ./2A3DFD6CE4CDFA79246F560DEADF33C3FE1DD0AEA2273BFED26CF5BF99086A0E.rs '' 'error: internal compiler error: compiler/rustc_hir_typeck/src/method/confirm.rs:518:21: Foo was a subtype of &Foo but now is not?', 'error: internal compiler error: compiler/rustc_hir_typeck/src/method/confirm.rs:518:21: Foo was a subtype of &Foo but now is not?'
File: /tmp/icemaker/2A3DFD6CE4CDFA79246F560DEADF33C3FE1DD0AEA2273BFED26CF5BF99086A0E.rs
-->

auto-reduced (treereduce-rust):
````rust
struct Foo(u32);
impl Foo {
    fn get<R>(self: R) -> u32 {
        self.0
    }
}

fn main() {
    let mut foo = Foo(1);
    foo.get::<&Foo>();
}


````

original:
````rust
#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord,
         Debug, Default, Copy)]

use std::ops::Deref;

struct Foo(u32);
impl Foo {
    fn get<R: Deref<Target=Self>>(self: R) -> u32 {
        self.0
    }
}

fn main() {
    let mut foo = Foo(1);
    foo.get::<&Foo>();
    //~^ ERROR mismatched types
}

````

Version information
````
rustc 1.75.0-nightly (bcb5798dd 2023-10-30)
binary: rustc
commit-hash: bcb5798dd890a691644af9d371f3bd7fcc465584
commit-date: 2023-10-30
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.3

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0307]: invalid `self` parameter type: R
 --> /tmp/icemaker_global_tempdir.UiXGlnwFiTMm/rustc_testrunner_tmpdir_reporting.qpHEuYT5OkHH/mvce.rs:3:21
  |
3 |     fn get<R>(self: R) -> u32 {
  |                     ^
  |
  = note: type of `self` must be `Self` or a type that dereferences to it
  = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)

error[E0609]: no field `0` on type `R`
 --> /tmp/icemaker_global_tempdir.UiXGlnwFiTMm/rustc_testrunner_tmpdir_reporting.qpHEuYT5OkHH/mvce.rs:4:14
  |
3 |     fn get<R>(self: R) -> u32 {
  |            - type parameter 'R' declared here
4 |         self.0
  |              ^

error: internal compiler error: compiler/rustc_hir_typeck/src/method/confirm.rs:518:21: Foo was a subtype of &Foo but now is not?
  --> /tmp/icemaker_global_tempdir.UiXGlnwFiTMm/rustc_testrunner_tmpdir_reporting.qpHEuYT5OkHH/mvce.rs:10:9
   |
10 |     foo.get::<&Foo>();
   |         ^^^

thread 'rustc' panicked at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/compiler/rustc_errors/src/lib.rs:1000:33:
Box<dyn Any>
stack backtrace:
   0:     0x7fbd04169cfc - std::backtrace_rs::backtrace::libunwind::trace::h696e701df900a2b2
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7fbd04169cfc - std::backtrace_rs::backtrace::trace_unsynchronized::h0a821b5067328428
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fbd04169cfc - std::sys_common::backtrace::_print_fmt::h561ba93d7dd378a2
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7fbd04169cfc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hadaea2aacdfa4b18
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7fbd041cbf50 - core::fmt::rt::Argument::fmt::haaf2e1dae680d0d4
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/core/src/fmt/rt.rs:142:9
   5:     0x7fbd041cbf50 - core::fmt::write::h92a090160f936a3a
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/core/src/fmt/mod.rs:1117:17
   6:     0x7fbd0415db8f - std::io::Write::write_fmt::h304b2c71e3a57237
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/io/mod.rs:1763:15
   7:     0x7fbd04169ae4 - std::sys_common::backtrace::_print::h805cb7447553b9fa
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7fbd04169ae4 - std::sys_common::backtrace::print::h7be441b9aecf7507
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7fbd0416c777 - std::panicking::default_hook::{{closure}}::hf0f1687c23b85f00
  10:     0x7fbd0416c4df - std::panicking::default_hook::h1aa7ed866efb6ff4
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/panicking.rs:292:9
  11:     0x7fbd06ce0b80 - std[52b4fac2c1cd6916]::panicking::update_hook::<alloc[b6c0ed09a4f055cf]::boxed::Box<rustc_driver_impl[f6d8c9ab4fbb7a4c]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7fbd0416ceb8 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::ha095ea6a5339338c
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/alloc/src/boxed.rs:2021:9
  13:     0x7fbd0416ceb8 - std::panicking::rust_panic_with_hook::h96c03960eff56a0b
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/panicking.rs:783:13
  14:     0x7fbd06e3ecb4 - std[52b4fac2c1cd6916]::panicking::begin_panic::<rustc_errors[56af66a7435c2564]::ExplicitBug>::{closure#0}
  15:     0x7fbd06e2b3a6 - std[52b4fac2c1cd6916]::sys_common::backtrace::__rust_end_short_backtrace::<std[52b4fac2c1cd6916]::panicking::begin_panic<rustc_errors[56af66a7435c2564]::ExplicitBug>::{closure#0}, !>
  16:     0x7fbd06e1e3e6 - std[52b4fac2c1cd6916]::panicking::begin_panic::<rustc_errors[56af66a7435c2564]::ExplicitBug>
  17:     0x7fbd06e11ffe - <rustc_errors[56af66a7435c2564]::HandlerInner>::span_bug::<rustc_span[219ab52d13232863]::span_encoding::Span, alloc[b6c0ed09a4f055cf]::string::String>
  18:     0x7fbd06e11d3b - <rustc_errors[56af66a7435c2564]::Handler>::span_bug::<rustc_span[219ab52d13232863]::span_encoding::Span, alloc[b6c0ed09a4f055cf]::string::String>
  19:     0x7fbd06e3eefb - rustc_middle[13e6374ce232bd81]::util::bug::opt_span_bug_fmt::<rustc_span[219ab52d13232863]::span_encoding::Span>::{closure#0}
  20:     0x7fbd06e41d0a - rustc_middle[13e6374ce232bd81]::ty::context::tls::with_opt::<rustc_middle[13e6374ce232bd81]::util::bug::opt_span_bug_fmt<rustc_span[219ab52d13232863]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7fbd06e32868 - rustc_middle[13e6374ce232bd81]::ty::context::tls::with_context_opt::<rustc_middle[13e6374ce232bd81]::ty::context::tls::with_opt<rustc_middle[13e6374ce232bd81]::util::bug::opt_span_bug_fmt<rustc_span[219ab52d13232863]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7fbd06056834 - rustc_middle[13e6374ce232bd81]::util::bug::span_bug_fmt::<rustc_span[219ab52d13232863]::span_encoding::Span>
  23:     0x7fbd08a3bd54 - <rustc_hir_typeck[8394991fd8d3c9ab]::method::confirm::ConfirmContext>::confirm
  24:     0x7fbd08a3140f - <rustc_hir_typeck[8394991fd8d3c9ab]::fn_ctxt::FnCtxt>::lookup_method
  25:     0x7fbd0884b3bc - <rustc_hir_typeck[8394991fd8d3c9ab]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  26:     0x7fbd083970d9 - <rustc_hir_typeck[8394991fd8d3c9ab]::fn_ctxt::FnCtxt>::check_block_with_expected
  27:     0x7fbd0884b2f0 - <rustc_hir_typeck[8394991fd8d3c9ab]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  28:     0x7fbd086b3a3f - rustc_hir_typeck[8394991fd8d3c9ab]::check::check_fn
  29:     0x7fbd08134354 - rustc_hir_typeck[8394991fd8d3c9ab]::typeck
  30:     0x7fbd081337c7 - rustc_query_impl[d72368bb483549e]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d72368bb483549e]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[13e6374ce232bd81]::query::erase::Erased<[u8; 8usize]>>
  31:     0x7fbd07f4aa2c - rustc_query_system[7f1cfd16d8bae603]::query::plumbing::try_execute_query::<rustc_query_impl[d72368bb483549e]::DynamicConfig<rustc_query_system[7f1cfd16d8bae603]::query::caches::VecCache<rustc_span[219ab52d13232863]::def_id::LocalDefId, rustc_middle[13e6374ce232bd81]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[d72368bb483549e]::plumbing::QueryCtxt, false>
  32:     0x7fbd07f4a710 - rustc_query_impl[d72368bb483549e]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7fbd083226d9 - rustc_hir_analysis[1263226015b2b1b2]::check_crate
  34:     0x7fbd08612796 - rustc_interface[6394f483f79c51a]::passes::analysis
  35:     0x7fbd086121b5 - rustc_query_impl[d72368bb483549e]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d72368bb483549e]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[13e6374ce232bd81]::query::erase::Erased<[u8; 1usize]>>
  36:     0x7fbd089231e6 - rustc_query_system[7f1cfd16d8bae603]::query::plumbing::try_execute_query::<rustc_query_impl[d72368bb483549e]::DynamicConfig<rustc_query_system[7f1cfd16d8bae603]::query::caches::SingleCache<rustc_middle[13e6374ce232bd81]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[d72368bb483549e]::plumbing::QueryCtxt, false>
  37:     0x7fbd08923015 - rustc_query_impl[d72368bb483549e]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  38:     0x7fbd08b1aeb4 - std[52b4fac2c1cd6916]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[6394f483f79c51a]::util::run_in_thread_with_globals<rustc_interface[6394f483f79c51a]::interface::run_compiler<core[68a6506991500897]::result::Result<(), rustc_span[219ab52d13232863]::ErrorGuaranteed>, rustc_driver_impl[f6d8c9ab4fbb7a4c]::run_compiler::{closure#1}>::{closure#0}, core[68a6506991500897]::result::Result<(), rustc_span[219ab52d13232863]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[68a6506991500897]::result::Result<(), rustc_span[219ab52d13232863]::ErrorGuaranteed>>
  39:     0x7fbd08b1a173 - <<std[52b4fac2c1cd6916]::thread::Builder>::spawn_unchecked_<rustc_interface[6394f483f79c51a]::util::run_in_thread_with_globals<rustc_interface[6394f483f79c51a]::interface::run_compiler<core[68a6506991500897]::result::Result<(), rustc_span[219ab52d13232863]::ErrorGuaranteed>, rustc_driver_impl[f6d8c9ab4fbb7a4c]::run_compiler::{closure#1}>::{closure#0}, core[68a6506991500897]::result::Result<(), rustc_span[219ab52d13232863]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[68a6506991500897]::result::Result<(), rustc_span[219ab52d13232863]::ErrorGuaranteed>>::{closure#1} as core[68a6506991500897]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  40:     0x7fbd04177c85 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5c45b6a24e5f1474
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/alloc/src/boxed.rs:2007:9
  41:     0x7fbd04177c85 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h18b35e832435abd8
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/alloc/src/boxed.rs:2007:9
  42:     0x7fbd04177c85 - std::sys::unix::thread::Thread::new::thread_start::h4880d4dbe7e5215f
                               at /rustc/bcb5798dd890a691644af9d371f3bd7fcc465584/library/std/src/sys/unix/thread.rs:108:17
  43:     0x7fbd03f419eb - <unknown>
  44:     0x7fbd03fc57cc - <unknown>
  45:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.75.0-nightly (bcb5798dd 2023-10-30) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `main`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0307, E0609.
For more information about an error, try `rustc --explain E0307`.

```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/check/wfcheck.rs
+++ b/compiler/rustc_hir_analysis/src/check/wfcheck.rs
@@ -94,7 +94,7 @@ pub(super) fn enter_wf_checking_ctxt<'tcx, F>(
     f: F,
 ) -> Result<(), ErrorGuaranteed>
 where
-    F: for<'a> FnOnce(&WfCheckingCtxt<'a, 'tcx>),
+    F: for<'a> FnOnce(&WfCheckingCtxt<'a, 'tcx>) -> Result<(), ErrorGuaranteed>,
 {
     let param_env = tcx.param_env(body_def_id);
     let infcx = &tcx.infer_ctxt().build();
@@ -105,7 +105,7 @@ where
     if !tcx.features().trivial_bounds {
         wfcx.check_false_global_bounds()
     }
-    f(&mut wfcx);
+    f(&mut wfcx)?;
 
     let assumed_wf_types = wfcx.ocx.assumed_wf_types_and_report_errors(param_env, body_def_id)?;
 
@@ -875,6 +875,7 @@ fn check_param_wf(tcx: TyCtxt<'_>, param: &hir::GenericParam<'_>) -> Result<(), 
                         ty,
                         trait_def_id,
                     );
+                    Ok(())
                 })
             } else {
                 let mut diag = match ty.kind() {
@@ -961,6 +962,7 @@ fn check_associated_item(
                 let ty = tcx.type_of(item.def_id).instantiate_identity();
                 let ty = wfcx.normalize(span, Some(WellFormedLoc::Ty(item_id)), ty);
                 wfcx.register_wf_obligation(span, loc, ty.into());
+                Ok(())
             }
             ty::AssocKind::Fn => {
                 let sig = tcx.fn_sig(item.def_id).instantiate_identity();
@@ -972,7 +974,7 @@ fn check_associated_item(
                     hir_sig.decl,
                     item.def_id.expect_local(),
                 );
-                check_method_receiver(wfcx, hir_sig, item, self_ty);
+                check_method_receiver(wfcx, hir_sig, item, self_ty)
             }
             ty::AssocKind::Type => {
                 if let ty::AssocItemContainer::TraitContainer = item.container {
@@ -983,6 +985,7 @@ fn check_associated_item(
                     let ty = wfcx.normalize(span, Some(WellFormedLoc::Ty(item_id)), ty);
                     wfcx.register_wf_obligation(span, loc, ty.into());
                 }
+                Ok(())
             }
         }
     })
@@ -1097,6 +1100,7 @@ fn check_type_defn<'tcx>(
         }
 
         check_where_clauses(wfcx, item.span, item.owner_id.def_id);
+        Ok(())
     })
 }
 
@@ -1121,7 +1125,8 @@ fn check_trait(tcx: TyCtxt<'_>, item: &hir::Item<'_>) -> Result<(), ErrorGuarant
     }
 
     let res = enter_wf_checking_ctxt(tcx, item.span, def_id, |wfcx| {
-        check_where_clauses(wfcx, item.span, def_id)
+        check_where_clauses(wfcx, item.span, def_id);
+        Ok(())
     });
 
     // Only check traits, don't check trait aliases
@@ -1164,6 +1169,7 @@ fn check_item_fn(
     enter_wf_checking_ctxt(tcx, span, def_id, |wfcx| {
         let sig = tcx.fn_sig(def_id).instantiate_identity();
         check_fn_or_method(wfcx, ident.span, sig, decl, def_id);
+        Ok(())
     })
 }
 
@@ -1218,6 +1224,7 @@ fn check_item_type(
                 tcx.require_lang_item(LangItem::Sync, Some(ty_span)),
             );
         }
+        Ok(())
     })
 }
 
@@ -1276,6 +1283,7 @@ fn check_impl<'tcx>(
         }
 
         check_where_clauses(wfcx, item.span, item.owner_id.def_id);
+        Ok(())
     })
 }
 
@@ -1548,11 +1556,11 @@ fn check_method_receiver<'tcx>(
     fn_sig: &hir::FnSig<'_>,
     method: ty::AssocItem,
     self_ty: Ty<'tcx>,
-) {
+) -> Result<(), ErrorGuaranteed> {
     let tcx = wfcx.tcx();
 
     if !method.fn_has_self_parameter {
-        return;
+        return Ok(());
     }
 
     let span = fn_sig.decl.inputs[0].span;
@@ -1571,11 +1579,11 @@ fn check_method_receiver<'tcx>(
     if tcx.features().arbitrary_self_types {
         if !receiver_is_valid(wfcx, span, receiver_ty, self_ty, true) {
             // Report error; `arbitrary_self_types` was enabled.
-            e0307(tcx, span, receiver_ty);
+            return Err(e0307(tcx, span, receiver_ty));
         }
     } else {
         if !receiver_is_valid(wfcx, span, receiver_ty, self_ty, false) {
-            if receiver_is_valid(wfcx, span, receiver_ty, self_ty, true) {
+            return Err(if receiver_is_valid(wfcx, span, receiver_ty, self_ty, true) {
                 // Report error; would have worked with `arbitrary_self_types`.
                 feature_err(
                     &tcx.sess.parse_sess,
@@ -1587,16 +1595,17 @@ fn check_method_receiver<'tcx>(
                     ),
                 )
                 .help(HELP_FOR_SELF_TYPE)
-                .emit();
+                .emit()
             } else {
                 // Report error; would not have worked with `arbitrary_self_types`.
-                e0307(tcx, span, receiver_ty);
-            }
+                e0307(tcx, span, receiver_ty)
+            });
         }
     }
+    Ok(())
 }
 
-fn e0307(tcx: TyCtxt<'_>, span: Span, receiver_ty: Ty<'_>) {
+fn e0307(tcx: TyCtxt<'_>, span: Span, receiver_ty: Ty<'_>) -> ErrorGuaranteed {
     struct_span_err!(
         tcx.sess.diagnostic(),
         span,
@@ -1605,7 +1614,7 @@ fn e0307(tcx: TyCtxt<'_>, span: Span, receiver_ty: Ty<'_>) {
     )
     .note("type of `self` must be `Self` or a type that dereferences to it")
     .help(HELP_FOR_SELF_TYPE)
-    .emit();
+    .emit()
 }
 
 /// Returns whether `receiver_ty` would be considered a valid receiver type for `self_ty`. If

--- a/tests/ui/self/arbitrary-self-from-method-substs.default.stderr
+++ b/tests/ui/self/arbitrary-self-from-method-substs.default.stderr
@@ -1,0 +1,13 @@
+error[E0658]: `R` cannot be used as the type of `self` without the `arbitrary_self_types` feature
+  --> $DIR/arbitrary-self-from-method-substs.rs:8:43
+   |
+LL |     fn get<R: Deref<Target = Self>>(self: R) -> u32 {
+   |                                           ^
+   |
+   = note: see issue #44874 <https://github.com/rust-lang/rust/issues/44874> for more information
+   = help: add `#![feature(arbitrary_self_types)]` to the crate attributes to enable
+   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0658`.

--- a/tests/ui/self/arbitrary-self-from-method-substs.feature.stderr
+++ b/tests/ui/self/arbitrary-self-from-method-substs.feature.stderr
@@ -1,5 +1,5 @@
 error[E0308]: mismatched types
-  --> $DIR/arbitrary-self-from-method-substs.rs:14:5
+  --> $DIR/arbitrary-self-from-method-substs.rs:16:5
    |
 LL |     foo.get::<&Foo>();
    |     ^^^ expected `&Foo`, found `Foo`

--- a/tests/ui/self/arbitrary-self-from-method-substs.rs
+++ b/tests/ui/self/arbitrary-self-from-method-substs.rs
@@ -1,10 +1,12 @@
-#![feature(arbitrary_self_types)]
+// revisions: default feature
+#![cfg_attr(feature, feature(arbitrary_self_types))]
 
 use std::ops::Deref;
 
 struct Foo(u32);
 impl Foo {
-    fn get<R: Deref<Target=Self>>(self: R) -> u32 {
+    fn get<R: Deref<Target = Self>>(self: R) -> u32 {
+        //[default]~^ ERROR: `R` cannot be used as the type of `self`
         self.0
     }
 }
@@ -12,5 +14,5 @@ impl Foo {
 fn main() {
     let mut foo = Foo(1);
     foo.get::<&Foo>();
-    //~^ ERROR mismatched types
+    //[feature]~^ ERROR mismatched types
 }

